examples/step-97/step-97.cc

```diff
@@ -0,0 +1,2170 @@
+      }
+    else
+      {
+        Bj = 0.5 * (pow(b2, 2) - pow(r, 2)) * F1 +
+             0.2 * (pow(r, 5) - pow(a2, 5)) * F2;
+      }
+
+    return Bj;
+  }
+
+  // This function computes the magnetic field induced by the magnetic core, see
+  // the expression for $\vec{B}_{\mu}$ in the introduction.
+  inline Tensor<1, 3> magnetic_field_core(double x,
+                                          double y,
+                                          double z,
+                                          double H0,
+                                          double mur,
+                                          double mu0,
+                                          double a1,
+                                          double b1)
+  {
+    double a3 = std::pow(a1, 3);
+    double b3 = std::pow(b1, 3);
+
+    double OMEGA   = ((mur - 1.0) / (mur + 2.0)) * (a3 / b3);
+    double gamma_1 = (-3.0 * b3 * H0 * OMEGA) /
+                     ((2.0 * mur + 1.0) - 2.0 * (mur - 1.0) * OMEGA);
+    double beta_1  = ((2.0 * mur + 1.0) * gamma_1) / ((mur - 1.0) * a3);
+    double alpha_1 = (-b3 * H0 + 2.0 * mur * gamma_1 - mur * b3 * beta_1) / 2.0;
+    double delta_1 = (mur * a3 * beta_1 - 2.0 * mur * gamma_1) / a3;
+
+    double r  = sqrt(pow(x, 2) + pow(y, 2) + pow(z, 2));
+    double r3 = std::pow(r, 3);
+    double r5 = std::pow(r, 5);
+
+    double zz = -3.0 * z * z / r5;
+    double xz = -3.0 * x * z / r5;
+    double yz = -3.0 * y * z / r5;
+
+    double mu;
+
+    Tensor<1, 3> grad_psi;
+    Tensor<1, 3> Bmu;
+    Tensor<1, 3> Happl;
+
+    if (r <= a1)
+      {
+        grad_psi[0] = 0.0;
+        grad_psi[1] = 0.0;
+        grad_psi[2] = delta_1;
+        mu          = mu0;
+      }
+    else if (r >= b1)
+      {
+        grad_psi[0] = alpha_1 * xz;
+        grad_psi[1] = alpha_1 * yz;
+        grad_psi[2] = -H0 + alpha_1 / r3 + alpha_1 * zz;
+        mu          = mu0;
+      }
+    else
+      {
+        grad_psi[0] = gamma_1 * xz;
+        grad_psi[1] = gamma_1 * yz;
+        grad_psi[2] = beta_1 + gamma_1 / r3 + gamma_1 * zz;
+        mu          = mur * mu0;
+      }
+
+    Happl[0] = 0.0;
+    Happl[1] = 0.0;
+    Happl[2] = H0;
+
+    Bmu = -mu * grad_psi - mu0 * Happl;
+
+    return Bmu;
+  }
+
+  // This class implements the closed-form analytical expression for the
+  // free-current density, $\vec{J}_f$, in the entire domain.
+  class FreeCurrentDensity : public Function<3>, public Settings
+  {
+  public:
+    FreeCurrentDensity()
+      : Function<3>(3)
+    {}
+
+    virtual void
+    vector_value_list(const std::vector<Point<3>> &p,
+                      std::vector<Vector<double>> &values) const override final
+    {
+      Assert(values.size() == p.size(),
+             ExcDimensionMismatch(values.size(), p.size()));
+
+      Tensor<1, 3> Jf;
+      double       r;
+
+      for (unsigned int i = 0; i < values.size(); i++)
+        {
+          r = sqrt(pow(p[i][0], 2) + pow(p[i][1], 2) + pow(p[i][2], 2));
```

Comment:
In general, qualify names like this:
```suggestion
          r = std::sqrt(std::pow(p[i][0], 2) + std::pow(p[i][1], 2) + std::pow(p[i][2], 2));
```
But in the specific case here, this is simpler:
```suggestion
          r = p[i].norm();
```